### ISSUE TITLE
Set-Content should not allow writing to multiple files using wildcards

### ISSUE THREAD

**Issue by mklement0** [PowerShell/PowerShell#6729]
<sup>Follow-up from #6714</sup>

Currently, `Set-Content` writes to whatever existing files match a wildcard pattern passed to its (implied) `-Path` parameter.

For instance, the following will overwrite _all_ existing `*.txt` files in the current directory:

```powershell
'hi' | Set-Content *.txt # !! Overwrites ALL *.txt files
```

This behavior is risky and should be changed to match the more sensible behavior of other output-to-file cmdlets such as `Out-File` (and therefore the `>` / `>>` operators), which generate an _error_ if the wildcard expression resolves to more than 1 file.

It is worth examining _all_ output-to-file cmdlets on this occasion.

Technically, this is a breaking change, but presumably one that falls into [Bucket 3: Unlikely Grey Area](https://github.com/PowerShell/PowerShell/blob/master/docs/dev-process/breaking-change-contract.md#bucket-3-unlikely-grey-area).

Environment data
----------------

```powershell
PowerShell Core v6.0.2 on macOS 10.13.4
PowerShell Core v6.0.2 on Ubuntu 16.04.4 LTS
PowerShell Core v6.0.2 on Microsoft Windows 10 Pro (64-bit; Version 1709, OS Build: 16299.371)
Windows PowerShell v5.1.16299.251 on Microsoft Windows 10 Pro (64-bit; Version 1709, OS Build: 16299.371)
```

**Comment by BrucePay** [comment]
Assuming they knew about it, I easily can see people using this to (re)initialize a set of file so I don't think it's really a bucket 3 change. Consider that
```
'hi' | Set-Content *.txt
```
is certainly much pithier than 
```
foreach ($f in dir *.txt) { "hi" | set-content $f } 
```

**Comment by mklement0** [comment]
Yes, as with any breaking change that takes away functionality there is a risk that people have relied on it.

My _guess_ is that, given that the behavior is (a) surprising to begin with and (b) is not present in `Out-File` / `>`, it's unlikely, but it's just a guess.

As for the pithiness; not quite as concise as the original, but a reasonable substitute, in my opinion, given the benefits of taking away the risky behavior:

```powershell
'hi' | Set-Content (Get-Item *.txt)
```

Note that, due to #6057, `(Get-Item *.txt).PSPath` must currently be used for full robustness.

**Comment by BrucePay** [comment]
Based on
```
'hi' | Set-Content (Get-Item *.txt)
```
your issue is not what I thought it was. It appears that you still want `Set-Content` to be able to write identical content to multiple files. Is this the case?

**Comment by mklement0** [comment]
> It appears that you still want Set-Content to be able to write identical content to multiple files. Is this the case?

Yes, if you pass an _array_ of _literal_ file paths to `-Path` / `-LiteralPath`, which to me unambiguously signals the intent.

_No change need there_, because these parameters are already typed `[string[]]`.

By contrast, the _wildcard_-based way of targeting multiple file is the fraught behavior, which I think should be eliminated.

I used the `'hi' | Set-Content (Get-Item *.txt)` example to demonstrate that someone who truly wants to target multiple files will still be able to do so, even once the wildcard-based method is taken away.

Now you could ask what should happen if an _array_ is passed to `-Path` (possibly implicitly) and at least one _element_ of that array is a wildcard expression?

For simplicity, my vote is for disallowing _any_ wildcard expression that matches more than 1 file, whether stand-alone or as part of an array.

To summarize, the desired behavior is:

* You can pass one or more _literal_ paths to `-LiteralPath`.

* You can pass one or more  literal paths or wildcard patterns - in any combination - to `-Path`, as long as no single wildcard pattern matches more than 1 file.

**Comment by vexx32** [comment]
This seems like the sort of thing that should be the default, but able to be worked around. You've provided one workaround, but consider perhaps that we could also allow the `-Force` parameter to bypass this restriction. In other words:
```powershell
# This doesn't work:
"blank" | Set-Content -Path "*.txt"

# But this does
"blank" | Set-Content -Path "*.txt" -Force
```

**Comment by mklement0** [comment]
@vexx32: 

Having it opt-in is definitely preferable, but note that we'd be adding meaning to the existing use of `-Force` ("Forces the cmdlet to set the contents of a file, even if the file is read-only.")

Also, you could argue that we'd have to implement the same thing for _all_ output-to-file cmdlets (`Out-File`, `Export-Csv`, ...)

On the other hand, as I've just realized, `Set-Content` already _is_ special in that at least at first glance it appears to be the only one that supports an _array_ of output filenames.

**Comment by iSazonov** [comment]
I think we could remove the dangerous behavior. In File System Provider at least. And follow to one principle for all cmdlet - target path parameter should always be LiteralPath. (We already did this for web cmdlets and Out-File is under question too #9475.)

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee looked at this and we compared it to behavior in `zsh` and `bash`.  It seems that `zsh` does support redirection to multiple files matching the wildcard while `bash` complains that the destination is ambiguous.  Also given that there seems to be legitimate cases where you want to initialize multiple files to the same content, then we think this is not a bucket 3 breaking change.  Also, if `-WhatIf` is used with `Set-Content` it does inform you of what was matched and would be written.  As noted, `Out-File` restricts to matching a single file so we have a bit of inconsistency between `Out-File` and `Set-Content`, but those two cmdlets don't have to be identical otherwise they'd be a single cmdlet.  We believe this is `By Design`

**Comment by iSazonov** [comment]
> As noted, `Out-File` restricts to matching a single file so we have a bit of inconsistency between `Out-File` and `Set-Content`, but those two cmdlets don't have to be identical otherwise they'd be a single cmdlet. We believe this is `By Design`

@SteveL-MSFT Please clarify. Out-File (and >,>>) uses `-FilePath` by default. This works like `-Path` with globbing but with returning only one path. And question is should we switch the default to LiteralPath in all scenarios (if we get path from pipeline we already bind to LiteralPath by PSPath).
